concepts/recursion/about.md

```diff
@@ -0,0 +1,191 @@
+                                     + [date(year, month, day_num + (ordinal - 1) * 7)])
+    return paydates_for_year_rec(year, weekday, ordinal, month, day_num + 1, output)
+
+    # find the second Wednesday of the month for all the months in 2022
+    print(paydates_for_year_rec(2022, 3, 2, 1, 1, []))
+    
+```
+
+Adya is happy that there are no more nested loops, no mutated state, and 2 fewer lines of code!  
+
+She is a little concerned that the recursive approach uses more steps than the looping approach, and so is less "performant".
+But re-writing the problem using recursion has definitely helped her deal with ugly nested looping (_a performance hazard_), extensive state mutation, and confusion around complex conditional logic.
+It also feels more "readable" - she is sure that when she comes back to this code after a break, she will be able to read through and remember what it does more easily. 
+
+In the future, Adya may try to work through problems recursively first.
+She may find it easier to initially walk through the problem in clear steps when nesting, mutation, and complexity are minimized.
+After working out the basic logic, she can then focus on optimizing her initial recursive steps into a more performant looping approach.
+
+Even later, when she learns about `tuples`, Adya could consider further "optimizing" approaches, such as using a `list comprehension` with `Calendar.itermonthdates`, or memoizing certain values.
+
+## Recursive Variation: The Tail Call
+
+A tail call is when the last statement of a function only calls itself and nothing more.
+This example is not a tail call, as the function adds 1 to the result of calling itself
+
+```python
+def print_increment(step, max_value):
+    if step > max_value:
+        return 1
+    print(f'The step is {step}')
+    return 1 + print_increment(step + 1, max_value)
+
+
+def main():
+    retval = print_increment(1, 2)
+    print(f'retval is {retval} after recursion')
+
+if __name__ == "__main__":
+    main()
+
+```
+
+This will print
+
+```
+The step is 1
+The step is 2
+retval is 3 after recursion
+```
+
+To refactor it to a tail call, make `j` a parameter of `printIncrement`
+
+```python
+def print_increment(step, max_value, retval):
+    if step > max_value:
+        return retval
+    print(f'The step is {step}')
+    return print_increment(step + 1, max_value, retval + 1)
+
+
+def main():
+    retval = print_increment(1, 2, 1)
+    print(f'retval is {retval} after recursion')
+
+if __name__ == "__main__":
+    main()
+
+```
+
+You may find a tail call even easier to reason through than a recursive call that is not a tail call.
+However, it is always important when using recursion to know that there will not be so many iterations that the stack will overflow.
+
+## Recursion Limits in Python
+
+Some languages are able to optimize tail calls so that each recursive call reuses the stack frame of the first call to the function, instead of adding another stack frame.
```

Comment:
```suggestion
Some languages are able to optimize tail calls so that each recursive call reuses the stack frame of the first call to the function (_similar to the way a loop reuses a frame_), instead of adding an additional frame to the stack.
```